### PR TITLE
Give imported packages lower precedence than standard library.

### PR DESCRIPTION
**Pull request by katzdm**
The current Bazel Python stub template prepends imported packages onto
`PYTHONPATH`. This effectively gives them higher `import`-precedence
than any directories:
- included in `PYTHONPATH` by the user
- included in `PYTHONPATH` or `sys.path` by the interpreter
The later category includes the Python standard library.

This can cause real and nontrivial problems for projects. If a project
uses uses `pip_import` to depend on `flake8`, then it will also
transitively import `enum34`, a library which provides compatibility
with the `enum` library (introduced in Python 3.4) for pre-3.4 versions
of Python, but expects never to be imported by Python>=3.4. The scheme
currently used by Bazel will find the `enum34` module prior to finding
the Standard Library `enum` module, even when run by Python3.7.

The solution proposed here is to build and load a `usercustomize` module
at program start, which will be loaded during program startup. This
module appends the directory of each imported module to the `sys.path`
variable, effectively making the packages available, but giving them
lower precedence than built-in directories.

cc: @brandjon

### CODE REVIEW

**Comment by Globegitter** [comment]
This sounds great and seems it could also fixe the issue we are running into with https://github.com/google/containerregistry/pull/109 - where https://github.com/agronholm/pythonfutures is used for python 2 futures compatibility but it does not work with python 3, so there it should import the syslib but it is giving preference to the dependency.

**Comment by katzdm** [comment]
Friendly ping - Jon, any thoughts as to whether a change like this could be workable?

**Comment by tmc** [comment]
This seems like a pretty important issue.

**Comment by tmc** [comment]
ping @brandjon - AFAICT this is part of improving python ecosystem support in bazel.

**Comment by katzdm** [comment]
Hi @brandjon and @aiuto - Can I ask for an update here? This PR was opened almost two months ago, and no feedback has been given. I understand if the change is more subtle/complex than it appears for google3-reasons, but I'd appreciate some indication of whether or not this (or something like it) is a viable direction.

Thanks in advance!

**Comment by brandjon** [comment]
Hi Daniel, apologies for the delay. I've fallen behind on triaging some Python PRs as a result of some feature work on the 2/3 mode issue. I'm on a release manager rotation this week but aim to work through this backlog in the coming days.

A quick comment about the substance of the change: Is it ever desirable in practice to continue to allow packages to override standard library modules? I.e., is that behavior consistent with any common workflow or default configuration in normal Python programming? (For example, ordinarily when there is a file whose name clashes with a standard library module, other files in the same directory will import it instead of the standard library. Though that particular example shouldn't be a concern for Bazel, since we use module names qualified by the package path.) If the answer is yes then we should provide some way to still get the old behavior, otherwise this is basically a bug fix and doesn't need to preserve compatibility.

**Comment by katzdm** [comment]
Hey Brandon - Thanks for getting back to me!

I'm a relative newcomer to the Python ecosystem myself, so I can't speak to whether there are idiomatic examples of overriding the standard library. I would imagine that this ought to be discouraged.

I believe I saw something about Bazel, in the future, allowing alternate template-stubs for Python "binaries". Perhaps this is the answer to projects that wish to use the legacy import-ordering? As you mentioned, it shouldn't be too much of an issue for Bazel, since it generally requires fully qualified package paths.

**Comment by katzdm** [comment]
What do you think of the following:
- Announce the intended change in #general and #python of BazelBuild Slack
- Send an email to the "Bazel/Python Special Interest Group" mailing list
- Reach out to some internal Google teams that might have more knowledge of whether this would break common idioms (Abseil Python might have some good perspective?)
- If we don't hear any concerns by end-of-week, then we move forward on Monday

Thoughts?

**Comment by brandjon** [comment]
Reviewed (see above) -- the PR sounds reasonable and it's just a matter of [confirming with the community](https://groups.google.com/forum/?nomobile=true#!forum/bazel-sig-python). I'm hoping to avoid the overhead of an incompatible change flag for this.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt [hunk 1]
```diff
@@ -148,11 +164,16 @@ def Main():
...
+  # directories to `sys.path`.
+  usercustomize_module = open("%s/usercustomize.py" % module_space, "w")
+  usercustomize_module.write(BuildUserCustomizeModule(python_path_entries))
+  usercustomize_module.close()
 
   old_python_path = os.environ.get('PYTHONPATH')
-  python_path = os.pathsep.join(python_path_entries)
-  if old_python_path:
-    python_path += os.pathsep + old_python_path
+  python_path = (old_python_path if old_python_path else "") + os.pathsep + module_space
```
[the remark this answers is not in the document]
Comment (katzdm) [reply src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt]: Thanks for the suggestion - Done.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt [hunk 2]
```diff
@@ -133,6 +138,17 @@ def RunfilesEnvvar(module_space):
...
+  return "\n".join([
+      "# AUTOGENERATED FILE. DO NOT EDIT.",
+      "import sys",
+      "\n",
+      "import_paths = [",
+  ] + ["    \"%s\"," % path for path in import_paths] + [
+      "]",
+      "sys.path.extend(import_paths)",
+  ])
+
```
[the remark this answers is not in the document]
Comment (katzdm) [reply src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt]: For sure; much better - Done.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt [hunk 3]
```diff
@@ -147,6 +152,17 @@ def RunfilesEnvvar(module_space):
 
   return (None, None)
 
+def BuildUserCustomizeModule(import_paths):
+  return """\
+# AUTOGENERATED FILE. DO NOT EDIT.
```
Comment (brandjon) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt]: Nit: Can you mention this is produced by the python launcher stub script? So anyone digging around later is saved a step.

**Comment by brandjon** [comment]
Also -- what's the rationale for emitting a usercustomize file as opposed to just sticking them directly on the `python_path` in the stub script? I'm not opposed, just curious.

**Comment by katzdm** [comment]
> Also -- what's the rationale for emitting a usercustomize file as opposed to just sticking them directly on the `python_path` in the stub script? I'm not opposed, just curious.

I think my rationale was that since Python provides a well-defined mechanism for configuring the `import` search path, it would be better to use the provided abstraction than to modify the path directly. The runtime logic for building the path is [somewhat nontrivial](https://github.com/python/cpython/blob/master/Doc/library/site.rst). Sticking to provided abstractions also (hopefully) reduces likelihood that we'll come into conflict with any behavioral changes that Python might introduce to `import`-path-building in the future (though it's worth mentioning that I don't know of any such changes coming down the pipeline).

**Comment by brandjon** [comment]
While there are indeed many moving parts for determining how an import is resolved, it's not clear to me why usercustomize is a better hook than PYTHONPATH itself. I think of them as just two different alternative mechanisms without one necessarily being more preferred than the other. Actually, I would think of usercustomize as more appropriate for a persistent configuration change that applies to all builds done by a given user, a la bashrc.

But I'm fine leaving this as-is. It's not hard to change it later if needed.

(I imagine it might conflict with real usercustomize modules installed on the system, but consulting those would be non-hermetic anyway.)

Mailing list results look pretty positive on just making this change. Will give it at least to later today before merging.

**Comment by katzdm** [comment]
> While there are indeed many moving parts for determining how an import is resolved, it's not clear to me why usercustomize is a better hook than PYTHONPATH itself. I think of them as just two different alternative mechanisms without one necessarily being more preferred than the other. Actually, I would think of usercustomize as more appropriate for a persistent configuration change that applies to all builds done by a given user, a la bashrc.
> 
> But I'm fine leaving this as-is. It's not hard to change it later if needed.
> 
> (I imagine it might conflict with real usercustomize modules installed on the system, but consulting those would be non-hermetic anyway.)
> 
> Mailing list results look pretty positive on just making this change. Will give it at least to later today before merging.

I believe it may also have had to do with `import`-precedence; IIRC, all of `PYTHONPATH` is prepended to the `import`-path with higher precedence than standard libraries, whereas `usercustomize` is appended with lower precedence.

**Comment by brandjon** [comment]
Ah, I assumed the standard libraries *were* on the `PYTHONPATH`, but that makes sense.

I have a commit to update the failing test in accordance with this behavior change -- can you [give access](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests) to the PR branch?

**Comment by katzdm** [comment]
> Ah, I assumed the standard libraries _were_ on the `PYTHONPATH`, but that makes sense.
> 
> I have a commit to update the failing test in accordance with this behavior change -- can you [give access](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests) to the PR branch?

I believe the "Allow edits from maintainers" option should already be enabled.

**Comment by brandjon** [comment]
My mistake, pushed.

**Comment by brandjon** [comment]
The failures look legit. Example failure for `//src/test/shell/bazel:bazel_windows_example_test`, function `test_native_python`:

```
Traceback (most recent call last):
  File "\\?\D:\temp\Bazel.runfiles_zl35ntyr\runfiles\main\examples\py_native\bin.py", line 3, in <module>
    from examples.py_native.lib import GetNumber
ModuleNotFoundError: No module named 'examples'
```

[Test source code](https://github.com/bazelbuild/bazel/blob/0377b0dbf397d66bfea393e76d20b3bc3e86afdd/src/test/shell/bazel/bazel_windows_example_test.sh#L219)

It's failing to find the module entirely, not shadowing the module with something else of the same name, so it can't just be the order of `PYTHONPATH`. Maybe `usercustomize` works differently under windows? Maybe there's already a usercustomize.py available in the windows environment, shadowing the one you're generating here? I wonder if [`site.addsitedir()`](https://docs.python.org/2/library/site.html#site.addsitedir) would work instead?

**Comment by Profpatsch** [comment]
I think the issue & reproduction I opened yesterday is relevant to this pull request: https://github.com/bazelbuild/bazel/issues/7051

**Comment by katzdm** [comment]
Yep; failure looks legit to me as well - I don't have access to a Windows box, so I'm going to need to iterate a bit on this commit, using your CI to help debug.

**Comment by rickeylev** [comment]
Just to confirm: It sounds like this fixes issue #5899?

**Comment by brandjon** [comment]
Yes, thanks for the crosslink.

**Comment by ali5h** [comment]
`usercustomise` works only when `ENABLE_USER_SITE=True`. That maybe the reason it fails on windows. I think appending the paths to `sys.path` might more resilient to different compilations of python

**Comment by brandjon** [comment]
See #7959 for other `sys.path` related problems. For this particular issue I think `sitecustomize.py` might be the way to go, but that may be redundant if we end up replacing the stub script with a wrapper module that runs in the same interpreter process as user code.

**Comment by elistevens** [comment]
I realize that this is a very old branch, but is my reading correct that this change will make it harder for builds to isolate themselves from whatever random packages might be installed on the system?

If so, this seems like a step backwards to me. I'd much rather bazel python targets declare all their dependencies and operate with the equivalent of `python -S`.

Imagine the situation where the user is on a system that has package `foo==1.0` installed on the system, but they need `foo==2.5.1` for their build to actually work. If the system `foo` comes first, the user is left without a way to get a working build short of modifying the system install (which could very well break something else that needed the `1.0` API).

**Comment by aiuto** [comment]
Ping to everyone. 
@brandjon - is this still a good idea?
@katzdm - are you going to fix the tests?

**Comment by katzdm** [comment]
I'm not likely to have time or energy to revisit this PR in the near future - Would recommend closing for the time being, and restarting discussion of whether there's a better solution if it's still an issue for other users.

**Comment by aiuto** [comment]
Thanks for that update.

**Pull request closed without merging**